github-actions: Switch to Intel ICX compiler.

We need to switch to the new Intel ICX compilers in our github-actions worker later this year.

```
icpc: remark #10441: The Intel(R) C++ Compiler Classic (ICC) is deprecated and will be removed from product release in the second half of 2023. The Intel(R) oneAPI DPC++/C++ Compiler (ICX) is the recommended compiler moving forward. Please transition to use this compiler. Use '-diag-disable=10441' to disable this message.
```

Useful resources for porting:

- [Porting Guide for ICC Users to DPCPP or ICX](https://www.intel.com/content/www/us/en/developer/articles/guide/porting-guide-for-icc-users-to-dpcpp-or-icx.html)
- [Compile and Run MPI Programs Using C++ with SYCL](https://www.intel.com/content/www/us/en/developer/articles/technical/compile-and-run-mpi-programs-using-dpcpp-language.html)
- [Build googletest with IntelLLVM compilers. ](https://github.com/google/googletest/commit/e032d57642d2a042b9445b2199277bb24fee2eb5)

I adjusted the github-worker and got deal.II configured with the IntelLLVM `icpx` compiler. Excerpt from `detailed.log`:
```
#        CMAKE_CXX_COMPILER:     IntelLLVM 2023.1.0 on platform Linux x86_64
#                                /opt/intel/oneapi/compiler/2023.1.0/linux/bin/icpx
```

However, compilation fails on the bundled package `muparser` with the following warning (treated as error):
```
/dealii/bundled/muparser_v2_3_3/src/muParserTest.cpp:1558:38: error: explicit comparison with infinity in fast floating point mode [-Werror,-Wtautological-constant-compare]
                                                bCloseEnough &= (fabs(fVal[i]) != numeric_limits<value_type>::infinity());
                                                                 ~~~~~~~~~~~~~ ^  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
```

In a different context, this issue has been discussed in https://github.com/intel/compute-benchmarks/issues/13. So what we could also try here is to either disable the `tautological-constant-compare` warning, or disable fast math for this particular file. If there is a workaround, we could also propose a patch for muparser.

@tamiko -- What do you think?

On my branch I temporarily disabled the `-Werror` flag to catch all warnings caused by the default `icpx` compiler. You can find the log [here](https://github.com/marcfehling/dealii/actions/runs/5182233223/jobs/9338776880).

---

We basically have three types of code snippets that we need to find a workaround for.

1. `icpx` complains about `std::isnan` and our [`numbers::is_nan()`](https://github.com/dealii/dealii/blob/00e6fe71c9c8cf0ce0ba1cab3bc069022e444889/include/deal.II/base/numbers.h#L529-L533) function:
```
dealii/include/deal.II/base/numbers.h:532:12: warning: explicit comparison with NaN in fast floating point mode [-Wtautological-constant-compare]
     return std::isnan(x);
            ^~~~~~~~~~~~~
```
Maybe we can solve it with something like the following, and replace all occurences of `std::isnan` with `numbers::is_nan`.
```
 inline bool 
 is_nan(const double x) 
 {
#ifdef SOME_INDICATOR_THAT_FINITE_MATH_IS_ON
    return false;
#else 
    return std::isnan(x); 
#endif
 } 
```

2. We check if something is infinity, e.g.
```
dealii/source/fe/mapping_q.cc:583:17: warning: explicit comparison with infinity in fast floating point mode [-Wtautological-constant-compare]
  if (p_unit[0] == std::numeric_limits<double>::infinity())
      ~~~~~~~~~ ^  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Since `std::isinf` doesn't seem to please `icpx`, we could try `!std::isfinite` (or `!numbers::is_finite).

Or we could write our own `numbers::is_inf` function with the same trick as above.

3. We check if something is smaller than infinity, e.g.
```
dealii/include/deal.II/lac/utilities.h:435:31: warning: explicit comparison with infinity in fast floating point mode [-Wtautological-constant-compare]
      const bool scale = (a_L < std::numeric_limits<double>::infinity());
                          ~~~ ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
We can use `std::isfinite` or (`numbers::is_finite`) here.


---

Here is a collection of code references that triggered warnings:

1. `is_nan`
https://github.com/dealii/dealii/blob/00e6fe71c9c8cf0ce0ba1cab3bc069022e444889/include/deal.II/base/numbers.h#L529-L533
https://github.com/dealii/dealii/blob/00e6fe71c9c8cf0ce0ba1cab3bc069022e444889/source/lac/solver_control.cc#L92
https://github.com/dealii/dealii/blob/00e6fe71c9c8cf0ce0ba1cab3bc069022e444889/include/deal.II/lac/full_matrix.templates.h#L1729
https://github.com/dealii/dealii/blob/00e6fe71c9c8cf0ce0ba1cab3bc069022e444889/source/lac/lapack_full_matrix.cc#L2464
https://github.com/dealii/dealii/blob/00e6fe71c9c8cf0ce0ba1cab3bc069022e444889/source/distributed/tria_base.cc#L616

2. `== infinity`
https://github.com/dealii/dealii/blob/00e6fe71c9c8cf0ce0ba1cab3bc069022e444889/source/fe/mapping_q.cc#L583
https://github.com/dealii/dealii/blob/00e6fe71c9c8cf0ce0ba1cab3bc069022e444889/source/fe/mapping_q.cc#L649
https://github.com/dealii/dealii/blob/00e6fe71c9c8cf0ce0ba1cab3bc069022e444889/source/particles/particle_handler.cc#L1264

3. `< infinity`
https://github.com/dealii/dealii/blob/00e6fe71c9c8cf0ce0ba1cab3bc069022e444889/include/deal.II/lac/utilities.h#L435

--- a/.github/workflows/linux.yml
+++ b/.github/workflows/linux.yml
@@ -330,8 +330,8 @@ jobs:
         mkdir build
         cd build
         cmake -D CMAKE_BUILD_TYPE=Debug \
-              -D CMAKE_CXX_COMPILER=icpc \
-              -D DEAL_II_CXX_FLAGS='-Werror' \
+              -D CMAKE_CXX_COMPILER=icpx \
+              -D DEAL_II_CXX_FLAGS='-Werror -fno-finite-math-only' \
               -D DEAL_II_EARLY_DEPRECATIONS=ON \
               -D DEAL_II_WITH_MPI=ON \
               -D DEAL_II_WITH_LAPACK=ON \

--- a/include/deal.II/grid/grid_reordering.h
+++ b/include/deal.II/grid/grid_reordering.h
@@ -35,7 +35,7 @@ DEAL_II_NAMESPACE_OPEN
  * this class. Usage of the old-style numbering is deprecated.
  */
 template <int dim, int spacedim = dim>
-class DEAL_II_DEPRECATED GridReordering
+class GridReordering
 {
 public:
   /**

--- a/include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h
+++ b/include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h
@@ -3857,7 +3857,7 @@ MGTwoLevelTransferNonNested<dim, LinearAlgebra::distributed::Vector<Number>>::
     // on element faces or vertices and therefore, the partitioner is fine
     IndexSet locally_relevant_dofs(dof_handler_fine.n_dofs());
     if (!this->fine_element_is_continuous &&
-        !dof_handler_fine.get_fe().degree == 0)
+        dof_handler_fine.get_fe().degree != 0)
       locally_relevant_dofs.add_indices(global_dof_indices.begin(),
                                         global_dof_indices.end());
 

--- a/source/base/qprojector.cc
+++ b/source/base/qprojector.cc
@@ -769,11 +769,11 @@ QProjector<2>::project_to_all_faces(const ReferenceCell &     reference_cell,
     {
       const auto support_points_line =
         [](const auto &face, const auto &orientation) -> std::vector<Point<2>> {
-        std::array<Point<2>, 2> vertices;
-        std::copy_n(face.first.begin(), face.first.size(), vertices.begin());
-        const auto temp =
-          ReferenceCells::Line.permute_according_orientation(vertices,
-                                                             orientation);
+        ArrayView<const Point<2>> vertices(face.first.begin(),
+                                           face.first.size());
+        const auto                temp =
+          ReferenceCells::Line.permute_by_combined_orientation(vertices,
+                                                               orientation);
         return std::vector<Point<2>>(temp.begin(),
                                      temp.begin() + face.first.size());
       };

--- a/source/base/qprojector.cc
+++ b/source/base/qprojector.cc
@@ -769,7 +769,8 @@ QProjector<2>::project_to_all_faces(const ReferenceCell &     reference_cell,
     {
       const auto support_points_line =
         [](const auto &face, const auto &orientation) -> std::vector<Point<2>> {
-        const Point<2> *          vertices_ptr = face.first.begin();
+        // MSVC struggles when using face.first.begin()
+        const Point<2, double> *  vertices_ptr = &face.first[0];
         ArrayView<const Point<2>> vertices(vertices_ptr, face.first.size());
         const auto                temp =
           ReferenceCells::Line.permute_by_combined_orientation(vertices,

--- a/source/base/qprojector.cc
+++ b/source/base/qprojector.cc
@@ -769,8 +769,8 @@ QProjector<2>::project_to_all_faces(const ReferenceCell &     reference_cell,
     {
       const auto support_points_line =
         [](const auto &face, const auto &orientation) -> std::vector<Point<2>> {
-        ArrayView<const Point<2>> vertices(face.first.begin(),
-                                           face.first.size());
+        const Point<2> *          vertices_ptr = face.first.begin();
+        ArrayView<const Point<2>> vertices(vertices_ptr, face.first.size());
         const auto                temp =
           ReferenceCells::Line.permute_by_combined_orientation(vertices,
                                                                orientation);

